res/controllers/Denon-HC1000S-scripts.js

```diff
@@ -0,0 +1,869 @@
+    } else {
+        if (engine.getValue(group, "loop_enabled")) {
+            engine.setValue(group, "loop_in_goto", 1); // to fix LoopIn hanging
+        } else {
+            engine.setValue(group, "loop_in", 1);
+        }
+    }
+};
+
+DenonHC1000S.loopOut = function(midino, control, value, status, group) {
+    print("ButtonStatusLoopOut triggered");
+    var shift = DenonHC1000S.shift;
+    if (shift) {
+        print("loopOut: shift is pressed");
+        engine.setValue(group, "loop_out", -1);
+        engine.setValue(group, "loop_end_position", -1);
+    } else {
+        if (value === 64) { //only on press not on release to avoid doubles
+            engine.setValue(group, "loop_end_position", -1); //removes current to fix hanging out
+            engine.setValue(group, "loop_out", 1); //and sets again
+            print("value is " + value);
+        }
+    }
+};
+
+DenonHC1000S.reloop = function(group, value, shift) {
+    var loopInPosition = engine.getValue(group, "loop_start_position");
+    var loopOutPosition = engine.getValue(group, "loop_end_position");
+    if (loopInPosition !== -1 && loopOutPosition !== -1) {
+        engine.setValue(group, "reloop_toggle", 1);
+    } else {
+        var samplesPerBeat = DenonHC1000S.samplesPerBeat(group);
+        loopInPosition = engine.getValue(group, "playposition") * engine.getValue(group, "track_samples");
+        if (shift) {
+            var loopInPositionInBeats = loopInPosition / samplesPerBeat;
+            if ((loopInPositionInBeats - Math.floor(loopInPositionInBeats)) > 0.5) {
+                loopInPosition = Math.floor(0.5 + loopInPositionInBeats) * samplesPerBeat;
+            } else {
+                loopInPosition = Math.floor(loopInPositionInBeats) * samplesPerBeat;
+            }
+        } else {
+            loopInPosition = engine.getValue(group, "playposition") * engine.getValue(group, "track_samples");
+        }
+        loopInPosition = Math.floor(0.5 + loopInPosition);
+        loopInPosition = Math.max(0, loopInPosition - loopInPosition % 2);
+        loopOutPosition = loopInPosition + Math.floor(0.5 + samplesPerBeat);
+        loopOutPosition = Math.max(0, loopOutPosition - loopOutPosition % 2);
+        if (loopInPosition + samplesPerBeat < engine.getValue(group, "track_samples")) {
+            engine.setValue(group, "loop_start_position", loopInPosition);
+            engine.setValue(group, "loop_end_position", loopOutPosition);
+            engine.setValue(group, "reloop_toggle", 1);
+        }
+    }
+};
+
+DenonHC1000S.resizeLoop = function(midino, control, value, status, group) {
+    var knobDelta = DenonHC1000S.getKnobDelta(value);
+    print(group + " loop_scale " + knobDelta);
+    var scaleFactor = 1;
+    if (!DenonHC1000S.shift) {
+        if (knobDelta > 0) { //make scale bigger (right turn)
+            if (!DenonHC1000S.resizeLoopButtonPressed) {
+                scaleFactor = (scaleFactor * scaleFactor + 1 / 2);
```

Comment:
I'm confused, this just scales the loop by `scaleFactor²+ 0.5`. If you're halving below, you probably would like to double the loop here using [`loop_double`](https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#control-[ChannelN]-loop_double).